File: backend/src/main/java/com/gridsync/crdt/HybridLogicalClock.java
package com.gridsync.crdt;

import java.time.Clock;
import java.util.UUID;

public record HybridLogicalClock(long physicalTime, int logicalCounter, UUID replicaId) implements Comparable<HybridLogicalClock> {

    private static long lastPhysicalTime = 0;
    private static int lastLogicalCounter = 0;
    private static final Object lock = new Object();

    public static HybridLogicalClock now(UUID replicaId, Clock clock) {
        synchronized (lock) {
            long currentPhysicalTime = clock.millis();
            if (currentPhysicalTime > lastPhysicalTime) {
                lastPhysicalTime = currentPhysicalTime;
                lastLogicalCounter = 0;
            } else {
                lastLogicalCounter++;
            }
            return new HybridLogicalClock(lastPhysicalTime, lastLogicalCounter, replicaId);
        }
    }

    // For testing purposes
    static void reset() {
        synchronized (lock) {
            lastPhysicalTime = 0;
            lastLogicalCounter = 0;
        }
    }

    @Override
    public int compareTo(HybridLogicalClock other) {
        int cmp = Long.compare(this.physicalTime, other.physicalTime);
        if (cmp != 0) return cmp;
        
        cmp = Integer.compare(this.logicalCounter, other.logicalCounter);
        if (cmp != 0) return cmp;
        
        if (this.replicaId != null && other.replicaId != null) {
            return this.replicaId.compareTo(other.replicaId);
        } else if (this.replicaId != null) {
            return 1;
        } else if (other.replicaId != null) {
            return -1;
        }
        return 0;
    }
}

File: frontend/src/crdt/HybridLogicalClock.ts
export class HybridLogicalClock {
  public readonly physicalTime: number;
  public readonly logicalCounter: number;
  public readonly replicaId: string;

  private static lastPhysicalTime: number = 0;
  private static lastLogicalCounter: number = 0;

  constructor(physicalTime: number, logicalCounter: number, replicaId: string) {
    this.physicalTime = physicalTime;
    this.logicalCounter = logicalCounter;
    this.replicaId = replicaId;
  }

  public static now(replicaId: string, customClock?: () => number): HybridLogicalClock {
    const currentPhysicalTime = customClock ? customClock() : Date.now();
    if (currentPhysicalTime > HybridLogicalClock.lastPhysicalTime) {
      HybridLogicalClock.lastPhysicalTime = currentPhysicalTime;
      HybridLogicalClock.lastLogicalCounter = 0;
    } else {
      HybridLogicalClock.lastLogicalCounter++;
    }
    return new HybridLogicalClock(
      HybridLogicalClock.lastPhysicalTime,
      HybridLogicalClock.lastLogicalCounter,
      replicaId
    );
  }

  // For testing
  public static reset() {
    HybridLogicalClock.lastPhysicalTime = 0;
    HybridLogicalClock.lastLogicalCounter = 0;
  }

  public compareTo(other: HybridLogicalClock): number {
    if (this.physicalTime !== other.physicalTime) {
      return this.physicalTime > other.physicalTime ? 1 : -1;
    }
    if (this.logicalCounter !== other.logicalCounter) {
      return this.logicalCounter > other.logicalCounter ? 1 : -1;
    }
    if (this.replicaId && other.replicaId) {
      return this.replicaId > other.replicaId ? 1 : (this.replicaId < other.replicaId ? -1 : 0);
    } else if (this.replicaId) {
      return 1;
    } else if (other.replicaId) {
      return -1;
    }
    return 0;
  }
}
